src/server.c

```diff
@@ -5532,13 +5564,81 @@ int prepareForShutdown(int flags) {
     if (server.loading || server.sentinel_mode)
         flags = (flags & ~SHUTDOWN_SAVE) | SHUTDOWN_NOSAVE;
 
-    int save = flags & SHUTDOWN_SAVE;
-    int nosave = flags & SHUTDOWN_NOSAVE;
+    server.shutdown_flags = flags;
 
     serverLog(LL_WARNING,"User requested shutdown...");
     if (server.supervised_mode == SUPERVISED_SYSTEMD)
         redisCommunicateSystemd("STOPPING=1\n");
 
+    /* If we have any replicas, let them catch up the replication offset before
+     * we shut down, to avoid data loss. */
+    if ((flags & SHUTDOWN_WAIT_REPL) &&
+        server.shutdown_timeout != 0 &&
+        !isReadyToShutdown())
+    {
+        server.shutdown_mstime = server.mstime + server.shutdown_timeout * 1000;
+        mstime_t cron_delay = 1000/server.hz;
+        pauseClients(server.shutdown_mstime + cron_delay, CLIENT_PAUSE_WRITE);
+        serverLog(LL_NOTICE, "Waiting for replicas before shutting down.");
+        return C_ERR;
+    }
+
+    return finishShutdown();
+}
+
+static inline int isShutdownInitiated(void) {
+    return server.shutdown_mstime != 0;
+}
+
+/* Returns 0 if there are any replicas which are lagging in replication which we
+ * need to wait for before shutting down. Returns 1 if we're ready to shut
+ * down now. */
+int isReadyToShutdown(void) {
+    if (server.masterhost != NULL) return 1; /* We're a replica. */
```

Comment:
@oranagra Do we want this line? Ideally we'd want a replica wait for sub-replicas to catch up. There might be more things to check for this case though. WDYT?

Reply:
i think you're right, sub-replicas should also wait.
by more things to check you mean write another 100 line test?

Reply:
No, I meant that if the master has exited, there is no way that the replica can catch up with the master and then it's no point letting it wait for sub-replicas to catch up because it will never happen.

    master -> replica -> sub-replica

So, maybe we want to stop waiting if we lost connection with the master? or something like that

    if (server.master_host != NULL && server.master == NULL) return 1; /* I'm a replica and I lost my master */

Reply:
for the point of that discussion i don't care about the master.
i think as a middle replica (one that has a sub-replica), when we shut down, we only care about data we have, which we didn't yet pass to our sub-replica.

Reply:
OK. I'll just delete this line then.

Reply:
In a middle-replica, I suppose `server.master_repl_offset` is the middle-replicas offset, not the actual master's offset. (The middle-replica acts as master towards the sub-prelica. Thus my slight confusion.)